chore(group): enforce lower case for item groups

sylvlecl opened the pull request:
More limited version of https://github.com/AntaresSimulatorTeam/AntaREST/pull/2283 
which will affect only groups, not cluster names and identifiers.

For v8.6 studies, short term storage groups are serialized to title case, in order
to comply with the stricter solver format at the time. 

Some improvements along the way:
- clarify versioning of commands:
  serialization format is versioned, not the commands themselves
- add and test some annotated type `ItemName` and `AreaId` that will be re-usable
- remove `group` from parent class `ItemProperties` which did not belong there

sylvlecl committed:
Backport group modifications from fix/lower-case-issues branch

Signed-off-by: Sylvain Leclerc <sylvain.leclerc@rte-france.com>

sylvlecl committed:
WIP

Signed-off-by: Sylvain Leclerc <sylvain.leclerc@rte-france.com>

sylvlecl committed:
WIP

Signed-off-by: Sylvain Leclerc <sylvain.leclerc@rte-france.com>

sylvlecl committed:
chore(commands): remove obsolete diff feature which hurt evolvability

Signed-off-by: Sylvain Leclerc <sylvain.leclerc@rte-france.com>

sylvlecl committed:
fix(tests): adapt ST storage tests

Signed-off-by: Sylvain Leclerc <sylvain.leclerc@rte-france.com>

sylvlecl committed:
fix(tests): fix mypy

Signed-off-by: Sylvain Leclerc <sylvain.leclerc@rte-france.com>

sylvlecl committed:
fix(tests): add test for parsing v1

Signed-off-by: Sylvain Leclerc <sylvain.leclerc@rte-france.com>

sylvlecl committed:
fix(tools): remove all variant management tools

Signed-off-by: Sylvain Leclerc <sylvain.leclerc@rte-france.com>

sylvlecl committed:
fix(tests): remove unused tests

Signed-off-by: Sylvain Leclerc <sylvain.leclerc@rte-france.com>

sylvlecl committed:
Merge branch 'dev' into fix/remove-command-diffs

sylvlecl committed:
fix(commands): remove last usage of command extractor

Signed-off-by: Sylvain Leclerc <sylvain.leclerc@rte-france.com>

sylvlecl committed:
Merge branch 'fix/remove-command-diffs' into fix/lower-case-group

Signed-off-by: Sylvain Leclerc <sylvain.leclerc@rte-france.com>

sylvlecl committed:
fix(tests): fix some tests in command factory

Signed-off-by: Sylvain Leclerc <sylvain.leclerc@rte-france.com>

sylvlecl committed:
fix(versioning): fix issues related to version handling in commands

Signed-off-by: Sylvain Leclerc <sylvain.leclerc@rte-france.com>

sylvlecl committed:
fix(tests): fix constraint on item names

Signed-off-by: Sylvain Leclerc <sylvain.leclerc@rte-france.com>

sylvlecl committed:
fix(tests): fix more tests

Signed-off-by: Sylvain Leclerc <sylvain.leclerc@rte-france.com>

sylvlecl committed:
fix(tests): fix last tests

Signed-off-by: Sylvain Leclerc <sylvain.leclerc@rte-france.com>

sylvlecl committed:
fix(clusters): remove unused property

Signed-off-by: Sylvain Leclerc <sylvain.leclerc@rte-france.com>

sylvlecl committed:
feat(ini): allow custom parsers/serializers for ini options

Signed-off-by: Sylvain Leclerc <sylvain.leclerc@rte-france.com>

sylvlecl committed:
fix(tests): improve tests

Signed-off-by: Sylvain Leclerc <sylvain.leclerc@rte-france.com>

sylvlecl committed:
fix(typing): stricter typing

Signed-off-by: Sylvain Leclerc <sylvain.leclerc@rte-france.com>

sylvlecl committed:
fix(typing): remove unions

Signed-off-by: Sylvain Leclerc <sylvain.leclerc@rte-france.com>

sylvlecl committed:
Merge remote-tracking branch 'origin/dev' into feat/custom-ini-serde

sylvlecl committed:
Merge remote-tracking branch 'origin/dev' into fix/lower-case-group

Signed-off-by: Sylvain Leclerc <sylvain.leclerc@rte-france.com>

sylvlecl committed:
Merge branch 'feat/custom-ini-serde' into fix/lower-case-group

sylvlecl committed:
feat(ini): parse and serialize group as lowercase

Signed-off-by: Sylvain Leclerc <sylvain.leclerc@rte-france.com>

MartinBelthle commented on antarest/study/business/areas/st_storage_management.py:
```diff
@@ -86,10 +89,11 @@ def validate_name(cls, name: t.Optional[str]) -> str:
-        values = self.model_dump(mode="json", by_alias=False, exclude_none=True)
-        return create_st_storage_config(study_version=study_version, **values)
+
+def _to_properties(version: StudyVersion, creation: STStorageCreation) -> STStoragePropertiesType:
```
Why didn't you put this as a method of STStorageCreation ? Because you did a method for RenewableCreation and ThermalCreation so we should probably harmonize

sylvlecl replied:
Moved it back to the class

MartinBelthle commented on antarest/study/storage/rawstudy/model/filesystem/config/field_validators.py:
```diff
@@ -9,12 +9,35 @@
+    if isinstance(name, int):
+        name = str(name)
+    if not isinstance(name, str):
+        ValueError(f"Invalid name '{name}'.")
```
Sonar Issue: We should add a `raise` here :)

MartinBelthle commented on antarest/study/storage/rawstudy/model/filesystem/root/input/st_storage/clusters/area/list.py:
```diff
@@ -10,10 +10,15 @@
 from antarest.study.storage.rawstudy.model.filesystem.context import ContextServer
 from antarest.study.storage.rawstudy.model.filesystem.ini_file_node import IniFileNode
 
+_VALUE_PARSERS = {any_section_option_matcher("group"): LOWER_CASE_PARSER}
```
Question: Should we put these parsers / serializers in a util file instead of declarating them every time ?

sylvlecl replied:
I think it's better to leave them here, each type of file must be responsible for its own parsing.
It would be an error to put this in a util file just to avoid a few duplication, because the format of each of the 3 files are not tied together and could evolve separately. Actually strictly speaking they are already different since we don't expect the same values in each one of those files, although we don't check it here, we could.

MartinBelthle commented on tests/integration/study_data_blueprint/test_renewable.py:
```diff
@@ -117,7 +117,7 @@ def test_lifecycle(
             **DEFAULT_PROPERTIES,
             "name": fr_solar_pv,
-            "group": "Solar PV",
+            "group": "solar pv",
```
I think here we should keep the MAJ to ensure we can create a cluster with a group in the wrong format.
We also have to change the check after to see that we return the value lowered

sylvlecl replied:
Done

MartinBelthle commented on antarest/study/storage/variantstudy/model/command/create_cluster.py:
```diff
@@ -42,50 +46,49 @@ class CreateCluster(ICommand):
+    def validate_model(cls, values: t.Dict[str, t.Any], info: ValidationInfo) -> t.Dict[str, t.Any]:
+        # Validate parameters
+        if isinstance(values["parameters"], dict):
+            parameters = copy.deepcopy(values["parameters"])
```
Don't think we need to make a copy here

MartinBelthle commented on antarest/study/storage/variantstudy/model/command/create_cluster.py:
```diff
@@ -42,50 +46,49 @@ class CreateCluster(ICommand):
+        if isinstance(values["parameters"], dict):
+            parameters = copy.deepcopy(values["parameters"])
+            if info.context and info.context.version == 1:
+                parameters["name"] = values["cluster_name"]
```
Could be done in one line: `parameters["name"] = values.pop("cluster_name")` as pop returns the value

MartinBelthle commented on tests/variantstudy/test_utils.py:
```diff
@@ -42,3 +44,24 @@ def test_aggregate_commands(command_context: CommandContext):
     assert len(command_dto_list) == 4
+
+
+def test_aggregate_commands_with_version(command_context: CommandContext):
```
Could be nice to test that 2 commands with different versions are not aggregated

sylvlecl replied:
Actually I removed this behaviour, it did not make sense because commands are not versioned, only DTOs are.
To make it clear, I removed `version` from the `ICommand` public API, and replaced it with a `_SERIALIZATION_VERSION` final field in implementations where it's needed.

We cannot have DTOs in different versions in this aggregation method, because they are all created from the same version of the code, so the test does not make sense anymore, I removed it.

MartinBelthle commented on antarest/study/storage/variantstudy/command_factory.py:
```diff
@@ -109,13 +116,15 @@ def _to_single_command(
-                version=version,
-                command_id=command_id,
-                study_version=study_version,
+            data = copy.deepcopy(args)
```
Same here, don't think you need a copy

sylvlecl replied:
Here I prefer to keep it, otherwise we will mutate the input DTO which is a weird side-effect. The arguments of the function will be modified by the function.

MartinBelthle replied:
Okay

MartinBelthle commented on antarest/study/storage/variantstudy/model/command/create_renewables_cluster.py:
```diff
@@ -39,21 +38,33 @@ class CreateRenewablesCluster(ICommand):
+    def validate_model(cls, values: t.Dict[str, t.Any], info: ValidationInfo) -> t.Dict[str, t.Any]:
+        # Validate parameters
+        if isinstance(values["parameters"], dict):
+            parameters = copy.deepcopy(values["parameters"])
```
Same remark on the copy and the one-line

MartinBelthle commented on antarest/study/storage/variantstudy/model/command/create_st_storage.py:
```diff
@@ -54,13 +59,18 @@ class CreateSTStorage(ICommand):
 
-    area_id: str = Field(description="Area ID", pattern=r"[a-z0-9_(),& -]+")
-    parameters: STStorageConfigType
+    # TODO SL: why lower case here and not in cluster / renewables ?
```
We can remove this TODO

sylvlecl replied:
Actually I left it because it is still relevant, just removed my name from it :)

sylvlecl replied:
Removed finally, see other comment

MartinBelthle commented on antarest/study/storage/variantstudy/model/command/create_st_storage.py:
```diff
@@ -54,13 +59,18 @@ class CreateSTStorage(ICommand):
-    area_id: str = Field(description="Area ID", pattern=r"[a-z0-9_(),& -]+")
-    parameters: STStorageConfigType
+    # TODO SL: why lower case here and not in cluster / renewables ?
+    area_id: LowerCaseStr = Field(description="Area ID", pattern=r"[a-z0-9_(),& -]+")
```
Do we want to keep the LowerCaseStr here ?

sylvlecl replied:
You're right, we should not change this in that PR, I will revert it and remove the TODO

sylvlecl replied:
I created an `AreaId` type used in the 2 other commands too, which also fixes the pattern constraint:
the pattern constraint does not work because it does not perform a full match, I added matching start and end of expression.

MartinBelthle commented on antarest/study/storage/variantstudy/model/command/create_st_storage.py:
```diff
@@ -85,13 +95,24 @@ class CreateSTStorage(ICommand):
+    @classmethod
+    def validate_model(cls, values: t.Dict[str, t.Any], info: ValidationInfo) -> t.Dict[str, t.Any]:
+        if isinstance(values["parameters"], dict):
+            parameters = copy.deepcopy(values["parameters"])
```
Same comment about the copy and the one-line

MartinBelthle commented on tests/integration/study_data_blueprint/test_st_storage.py:
```diff
@@ -361,7 +361,7 @@ def test_lifecycle__nominal(
         grand_maison_properties = {
             "name": grand_maison,
-            "group": "PSP_closed",
+            "group": "psp_closed",
```
Same. We could keep the old writing

sylvlecl replied:
Done

MartinBelthle commented on tests/integration/test_integration_token_end_to_end.py:
```diff
@@ -116,15 +116,15 @@ def test_nominal_case_of_an_api_user(client: TestClient, admin_access_token: str
                 "parameters": {
                     "group": "Gas",
                     "marginal-cost": 98,
-                    "unitCount": 1,
```
Did you change it because it doesn't work anymore ? If so, it could be an issue as it's kind of a breaking change (even though the scripts don't use this writing)

sylvlecl replied:
Actually on dev branch it just does not work, you get errors when the variant is generated, but we just don't check for them.
```
pydantic_core._pydantic_core.ValidationError: 1 validation error for ThermalConfig
unitCount
  Extra inputs are not permitted [type=extra_forbidden, input_value=1, input_type=int]
```

Now the input is validated as soon as the command is posted, that's why the test does not pass any more.

MartinBelthle commented on tests/storage/repository/filesystem/special_node/test_lower_case_nodes.py:
```diff
@@ -0,0 +1,118 @@
+
+    data = {"0": {"name": "BC_1", "group": "GRP_1"}}
+    node.save(data)
+    # Asserts the data is saved correctly
```
Could be nice to add the same test as above concerning the save of a new group specifically

sylvlecl replied:
Added. I also adapted the implementation and the test sor that we also *write* the group in lower case.

MartinBelthle commented on tests/variantstudy/model/command/test_create_cluster.py:
```diff
@@ -177,28 +176,63 @@ def test_to_dto(self, command_context: CommandContext):
-                "parameters": {"group": "Nuclear", "nominalcapacity": 2400, "unitcount": 2},
-                "prepro": prepro_id,
-                "modulation": modulation_id,
+                "modulation": "acc60faf9e0c47c77ec61c0d522382676102d8c1de1eb4aac1c5bb3a11682ff8",
```
Why did you change the prepro and modulation ids ? I think it was better before. Also this hash could be random  no ?

sylvlecl replied:
I reverted this change.

MartinBelthle commented on tests/variantstudy/test_command_factory.py:
```diff
@@ -27,401 +27,712 @@
+            args={
                 "name": "id",
-                "base_filter": "add-all",
+                "filter_items": ["a"],
```
Why did this change ?

sylvlecl replied:
It did not, it corresponds to line 46 before the PR (single district creation, not list creation).

MartinBelthle reviewed (changes requested):
As we discussed with Florian, we should do something in the run method in case the study is in version 8.6 or 8.7 and the user didn't specified a solver higher or equal to 8.8
Could be done in another PR but it would mean that this doesn't work on it's own

sylvlecl committed:
fix(test): add unit test for item name type

Signed-off-by: Sylvain Leclerc <sylvain.leclerc@rte-france.com>

sylvlecl committed:
fix(st-storage): fix consistency issue

Signed-off-by: Sylvain Leclerc <sylvain.leclerc@rte-france.com>

sylvlecl committed:
chore(commands): remove deepcopy and shorten code

Signed-off-by: Sylvain Leclerc <sylvain.leclerc@rte-france.com>

sylvlecl committed:
fix(validation): fix area id validation in commands

Signed-off-by: Sylvain Leclerc <sylvain.leclerc@rte-france.com>

sylvlecl committed:
chore(tests): change renewable integration test with upper case group

Signed-off-by: Sylvain Leclerc <sylvain.leclerc@rte-france.com>

sylvlecl committed:
chore(tests): restore upper case group in st storage test

Signed-off-by: Sylvain Leclerc <sylvain.leclerc@rte-france.com>

sylvlecl committed:
Merge remote-tracking branch 'origin/dev' into fix/lower-case-group

Signed-off-by: Sylvain Leclerc <sylvain.leclerc@rte-france.com>

sylvlecl committed:
chore(tests): restore computed matrix ids in test

Signed-off-by: Sylvain Leclerc <sylvain.leclerc@rte-france.com>

sylvlecl committed:
fix(tests): fix obsolete import of ini reader

Signed-off-by: Sylvain Leclerc <sylvain.leclerc@rte-france.com>

sylvlecl committed:
refacto(commands): remove version from public API of commands

Commands themselves are not versioned, only their serialization
format.
Also, the version must not be mutable, we make it
protected and final.

Signed-off-by: Sylvain Leclerc <sylvain.leclerc@rte-france.com>

sylvlecl committed:
fix(bc): improve test and write group to lower case

Signed-off-by: Sylvain Leclerc <sylvain.leclerc@rte-france.com>

sylvlecl committed:
fix(test): fix test

Signed-off-by: Sylvain Leclerc <sylvain.leclerc@rte-france.com>

MartinBelthle commented on antarest/study/storage/rawstudy/model/filesystem/config/field_validators.py:
```diff
@@ -9,12 +9,36 @@
 _ALL_FILTERING = ["hourly", "daily", "weekly", "monthly", "annual"]
 
 
+_VALID_NAME_PATTERN = re.compile(r"[a-zA-Z0-9_(),& -]+")
```
It seems you're not using this regex anywhere

MartinBelthle commented on antarest/study/storage/rawstudy/model/filesystem/config/field_validators.py:
```diff
@@ -9,12 +9,36 @@
+# does not comply with antares-simulator limitations.
+ItemName = Annotated[str, BeforeValidator(_validate_item_name)]
+
+AreaId = Annotated[str, Field(description="Area ID", pattern=r"^[a-z0-9_(),& -]+$")]
```
Not sure to understand why this pattern is different from the pattern used inside `transform_name_to_id`. Perhaps add a comment

sylvlecl replied:
This pattern is meant to be used as a full match, hence the `^` and `$`. Otherwise, the constraint does not do anything, because it partially matches more or less any string.
The characters inside brackets are valid characters for IDs.

In transform_name_to_id, the `^` is used to **negate** the characters in the bracket: we want to find characters that are **different** from valid characters.

sylvlecl replied:
Note: I've added some unit tests in test_field_validators to check the behaviour (before, the annotation in ST storage creation did not work properly)

MartinBelthle commented on antarest/study/storage/rawstudy/model/filesystem/config/st_storage.py:
```diff
@@ -160,7 +163,29 @@ class STStorage880Config(STStorage880Properties, LowerCaseIdentifier):
 # NOTE: In the following Union, it is important to place the older version first,
 # because otherwise, creating a short term storage always creates a v8.8 one.
-STStorageConfigType = STStorageConfig | STStorage880Config
+STStorageConfigType = Union[STStorageConfig, STStorage880Config]
```
You can replace `Union` by `|` to fit with a PR Theo merged recently

MartinBelthle commented on antarest/study/storage/rawstudy/model/filesystem/config/thermal.py:
```diff
@@ -408,7 +412,8 @@ class Thermal870Config(Thermal870Properties, IgnoreCaseIdentifier):
 # NOTE: In the following Union, it is important to place the most specific type first,
 # because the type matching generally occurs sequentially from left to right within the union.
-ThermalConfigType = Thermal870Config | Thermal860Config | ThermalConfig
+ThermalConfigType = Union[Thermal870Config, Thermal860Config, ThermalConfig]
```
Same comment about Union

MartinBelthle commented on antarest/study/storage/variantstudy/business/utils.py:
```diff
@@ -103,38 +103,58 @@ def decode(alias: str, study: FileStudy) -> str:
+    ref_command_dtos: Optional[Sequence[CommandDTO]] = None,
     force_aggregate: bool = False,
 ) -> List[CommandDTO]:
+    """
```
Nice to put a comment, I barely understand your "refactoring"

MartinBelthle commented on tests/variantstudy/test_utils.py:
```diff
@@ -9,10 +9,13 @@
 # SPDX-License-Identifier: MPL-2.0
 #
 # This file is part of the Antares project.
+import pytest
```
I think these imports are useless

MartinBelthle commented on antarest/study/storage/variantstudy/model/command/create_binding_constraint.py:
```diff
@@ -239,9 +241,7 @@ def to_dto(self) -> CommandDTO:
-        return CommandDTO(
-            action=self.command_name.value, args=args, version=self.version, study_version=self.study_version
-        )
+        return CommandDTO(action=self.command_name.value, args=args, version=1, study_version=self.study_version)
```
Perhaps instead of putting version=1 we should declare it as you did in `create_cluster`

sylvlecl replied:
Done

sylvlecl committed:
chore(misc): handle last review comments

Signed-off-by: Sylvain Leclerc <sylvain.leclerc@rte-france.com>

sylvlecl committed:
fix(sts): for v8.6 studies, STS groups are written in title case

Signed-off-by: Sylvain Leclerc <sylvain.leclerc@rte-france.com>

sylvlecl committed:
Merge branch 'dev' into fix/lower-case-group
